### ISSUE TITLE
A threaded message that is flagged with harassment or assault is not hidden from LHN and thread header

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#21497]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Send message from account A to account B
2. Switch to the tab where account B is signed in & reply to the message in thread
3. Navigate back to the sent message
4. Flag the message with harassment or assault

## Expected Result:
Flagged message shouldn't be displayed on thread header or LHN

## Actual Result:
Flagged message is displayed on thread header and LHN

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.29-0

**Reproducible in staging?:** y

**Reproducible in production?:** y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/0221cd2d-31d8-45b9-aee7-fbb2a47a4959

https://github.com/Expensify/App/assets/93399543/70a50761-657a-4792-ac5f-37a0ae6041ec

**Expensify/Expensify Issue URL:**

**Issue reported by:** @Natnael-guchima

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1686942349821779

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by AmjedNazzal** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

A threaded message that is flagged with harassment or assault is not hidden from LHN and thread header

### What is the root cause of that problem?

The text used for the titles of OptionRowLHN is retrieved from ReportUtils.getReportName which isn't handling the case where the parent message has been reported and hidden.
https://github.com/Expensify/App/blob/6183609c577f2310928bb4e8ee23b52f6dd28e00/src/libs/SidebarUtils.js#L372-L374

### What changes do you think we should make in order to solve the problem?

We could firstly add a new entry to parentReportAction and for example call it hiddenMessage and give it a value similar to deletedMessage.
We then can then add a check in ReportUtils.getReportName that would return hiddenMessage if the parentReport has MODERATOR_DECISION_PENDING_HIDE or MODERATOR_DECISION_PENDING_REMOVE.
```javascript
if (
    lodashGet(parentReportAction, 'message[0].moderationDecisions[0].decision') === CONST.MODERATION.MODERATOR_DECISION_PENDING_HIDE ||
    lodashGet(parentReportAction, 'message[0].moderationDecisions[0].decision') === CONST.MODERATION.MODERATOR_DECISION_PENDING_REMOVE
) {
    return Localize.translateLocal('parentReportAction.hiddenMessage');
}
```

### Result

https://github.com/Expensify/App/assets/74202751/0a26b4d4-6de5-463b-8fd1-c73566fe59d5

**Comment by StevenKKC** [comment]
### What is the root cause of that problem?
The thread report name fetched by `ReportUtils.getReportName` will be displayed on LHN and thread header.
https://github.com/Expensify/App/blob/6183609c577f2310928bb4e8ee23b52f6dd28e00/src/libs/ReportUtils.js#L996-L1037

In `getReportName`, there is no checking whether the threaded message is flaged, so the message is not hidden from LHN and thread header.

### What changes do you think we should make in order to solve the problem?
We can add checking the threaded message is or not flagged, and if flagged, then return suitable report name, ex. [Deleted message].
```
    if (isThread(report)) {
        ...
        if (parentReportAction.message[0].moderationDecisions) {
            const latestDecision = parentReportAction.message[0].moderationDecisions[0];
            if (latestDecision.decision === CONST.MODERATION.MODERATOR_DECISION_PENDING_REMOVE) {
                return Localize.translateLocal('parentReportAction.deletedMessage');
            }
        }
        ...
    }
```

### What alternative solutions did you explore? (Optional)
None.

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Flagging a thread message with Harrasment/Assault hides the message, but we can still see it in LHN.

### What is the root cause of that problem?
When we flag it as Harrasment/Assault, the moderation decision is `pendingRemove`.
https://github.com/Expensify/App/blob/820fa93ad1e799c12d8b639b98018b5350ea17af/src/libs/actions/Report.js#L1752-L1758

If the moderation decision is `pendingRemove`, we will not show it in the report screen.
https://github.com/Expensify/App/blob/820fa93ad1e799c12d8b639b98018b5350ea17af/src/libs/ReportActionsUtils.js#L328-L330

However, even though we hide it in the report screen, we don't hide it in LHN.
### What changes do you think we should make in order to solve the problem?

Hide the thread report if the thread message is flagged as Harrasment/Assault.

Details:
1. We have a code to decide whether to show a report in LHN.
https://github.com/Expensify/App/blob/820fa93ad1e799c12d8b639b98018b5350ea17af/src/libs/ReportUtils.js#L1901-L1924
In L1918, return `false` if the report is a thread and the thread parent message moderation decision is `pendingRemove`.
```
if (isThread(report) && lodashGet(ReportActionsUtils.getParentReportAction(report), 'message[0].moderationDecisions[0].decision') === CONST.MODERATION.MODERATOR_DECISION_PENDING_REMOVE) {
    return false;
}
```

2. We also need to show not found page if a user is trying to access it directly from a browser URL/deeplink
https://github.com/Expensify/App/blob/820fa93ad1e799c12d8b639b98018b5350ea17af/src/pages/home/ReportScreen.js#L239-L240
We just need to append it with the same condition as above.
```
const shouldHideReport = ... && (ReportUtils.isThread(this.props.report) && lodashGet(ReportActionsUtils.getParentReportAction(this.props.report), 'message[0].moderationDecisions[0].decision') === CONST.MODERATION.MODERATOR_DECISION_PENDING_REMOVE)
```

**Comment by Natnael-Guchima** [comment]
Here are the places where the flagged message is not hidden. 

![image-3.png](https://github.com/Expensify/App/assets/57330233/2b4a8226-1c0c-4acf-9802-2f9f6d247268)

**Comment by s77rt** [comment]
@AmjedNazzal Thanks for the proposal. You RCA is correct and the solution looks good to me.

:ribbon: :eyes: :ribbon: C+ reviewed

**Comment by s77rt** [comment]
@StevenKKC Thanks for the proposal. I think it's the same as @AmjedNazzal's proposal.

**Comment by s77rt** [comment]
@bernhardoj Thanks for the proposal. Your RCA makes sense but I don't think we want to hide the whole thread.

**Comment by bernhardoj** [comment]
>## Expected Result:
>Flagged message shouldn't be displayed on thread header or LHN

I thought we want to hide the message instead of renaming it to [Hidden]. If we don't hide it, user will be able to send more messages in that pendingRemove thread.

I think the correct use case of renaming it to [Hidden] is this issue https://github.com/Expensify/App/issues/21711

**Comment by AmjedNazzal** [comment]
@bernhardoj I think the idea here is not to stop the user from interacting with a thread while it's parent is pending remove but rather hiding the message in question until a decision has been reached.

**Comment by bernhardoj** [comment]
We allow a user to interact with the "pendingRemove" thread from LHN, but do not allow it from the chat/report itself. That doesn't sound consistent to me.

Anyway, I think we need confirmation of the expected behavior from the team.

**Comment by s77rt** [comment]
@bernhardoj 

> but do not allow it from the chat/report itself

Can you please elaborate on this? The report action is still rendered on the report

**Comment by bernhardoj** [comment]
If you flag the message with harassment or assault, the message will not be shown on the report/chat. (video >= 0:27)

**Comment by AmjedNazzal** [comment]
@bernhardoj This is not the expected behaviour suggested in the bug report though, the issue originally is that the supposedly hidden text is showing, not that the thread as a whole shoud be hidden.

**Comment by s77rt** [comment]
Sorry @AmjedNazzal I misinterpreted the bug, I didn't flag a new message because we were told not to flag until Wednesday. Now I see that those flagged messages are completely removed.

**Comment by AmjedNazzal** [comment]
@s77rt This wasn't the expected result in the original post, should I edit my proposal to account for this?

**Comment by AmjedNazzal** [comment]
Also what about when a message is pending hide and not remove?

**Comment by AmjedNazzal** [comment]
Furthermore not all flagged messages are removed, they can be pending remove but then not get removed if they are not actually harrasment or assult

**Comment by s77rt** [comment]
@AmjedNazzal The expected behaviour in the OP only gives us an idea of what the behaviour should be. That's from the perspective of the reporter. It can be exactly what we are looking as it can be different, this is usually something that we decide as a team when certain things are not clear. In the expected behaviour here it's stated that the flagged message should not be displayed in LHN. But it was not clear if we should show a different message instead of the flagged message (i.e. hide the message) or not show that flagged message at all (i.e. remove the message).

> Also what about when a message is pending hide and not remove?

For this case, your proposal is correct

> Furthermore not all flagged messages are removed, they can be pending remove but then not get removed if they are not actually harrasment or assult

Correct. Is there any problem that we are overlooking here?

**Comment by AmjedNazzal** [comment]
@s77rt I'm sorry but I just no longer understand the goal here, say someone flags a parent message, regardless if the flag is gonna return PENDING_HIDE or PENDING_REMOVE there will be a period of time where the report is proccessing, correct? during that proccessing period is what the bug reported seem to be about. so while the report is being proccessed, what is the expected behaviour? would the thread be completely hidden or would the heading and LHN text be replaced with "hidden"? also am I supposed to edit my proposal here or do I just move on to another issue?

**Comment by s77rt** [comment]
@AmjedNazzal If it's flagged as:
- Spam/Inconsiderate: Nothing will change. You can still view the message and access the thread (view replies). The thread can also be seen in LHN
- Intimidation/Bullying: The message will show as [Hidden]. You can still access the replies. The thread can be seen in LHN but titled [Hidden]
- Harassment/Assault: The message will no longer be visible (nothing is shown at all). You can't access the thread (replies). The thread should not be displayed in LHN at all


Recently flagging as Harassment/Assault behaved the same as flagging as Intimidation/Bullying so I misinterpreted the bug and the expected behaviour.

What are the next steps here? I think we have two plans:

**Plan A:** Change the title of this issue to handle both the to be hidden and to be removed cases. Assign @bernhardoj to fix both cases. Split the bounty 50% 50% (@bernhardoj and @AmjedNazzal) but @bernhardoj would be the one who raises the PR and the only eligible for speed bonus (in case merged within 3 days).

**Plan B:** Only handle the to be removed case. Assign @bernhardoj and move on normally. Report the to be hidden case in Slack and submit a proposal on the newly created issue.


@AmjedNazzal, @bernhardoj Let me know which plan you prefer here (cc @AndrewGable, @mallenexpensify)

**Comment by AmjedNazzal** [comment]
@s77rt In my opinion it makes more sense to split the cases as they are pretty different and I can just make a new bug report on slack regarding the the Intimidation/Bullying case and move my proposal to there with minor tweak to only work with PENDING_HIDE

**Comment by bernhardoj** [comment]
I agree with @AmjedNazzal. I think we need a separate discussion to discuss the expected behavior of "pendingHide" message.

**Comment by s77rt** [comment]
Super. Thank you! 

@AndrewGable Can you please assign @bernhardoj for his proposal here https://github.com/Expensify/App/issues/21497#issuecomment-1606100490.

**Comment by AmjedNazzal** [comment]
@s77rt can you please change the title of this post so it's not confusing for when I make a new bug report? I think it should be something like "A threaded message that is flagged with harassment or assault is not removed from LHN"

**Comment by AmjedNazzal** [comment]
Can the title of this please be changed to reflect the actual issue? already a post of mine got marked as dupe from this

**Comment by Natnael-Guchima** [comment]
@s77rt I have mentioned all the impacted areas here https://github.com/Expensify/App/issues/21497#issuecomment-1608973606. Incase, if it is not seen.

**Comment by s77rt** [comment]
@AndrewGable, @mallenexpensify or @greg-schroeder can you please update the title to: A threaded message that is flagged with harassment or assault is not removed from LHN

**Comment by bernhardoj** [comment]
PR is ready for review

cc: @s77rt

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: This existed since the implementation of threads, but the inconsistency was introduced by https://github.com/Expensify/App/pull/20223
- The offending PR has been commented on: https://github.com/Expensify/App/pull/20223#discussion_r1251827651
- A discussion in #expensify-bugs has been started: https://expensify.slack.com/archives/C049HHMV9SM/p1688467202306489
- Determine if we should create a regression test for this bug: I don't see any test suite/case for flag systems. I think we should add a regression test for this one.

___

**Regression Test Proposal**

Prerequisites: User B sends a message to User A
The below step is for User A

1. Press "Reply in thread" of the message sent by B
2. Send any message in the thread
3. Verify you see the new thread on LHN
4. Go back to the chat with B
5. Flag B's message as Harassment/Assault
6. Verify the message is removed
7. Verify the thread chat on LHN is removed

**Comment by greg-schroeder** [comment]
Processing payment now @Natnael-Guchima @s77rt @bernhardoj

**Comment by Natnael-Guchima** [comment]
@greg-schroeder I have accepted the offer. Thanks.

**Comment by greg-schroeder** [comment]
Regression test filed, all offers sent out. closing
